prow/spyglass/gcsartifact_fetcher.go

```diff
@@ -0,0 +1,256 @@
+	return &GCSJobSource{
+		source:     src,
+		linkPrefix: linkPrefix,
+		bucket:     bucket,
+		jobPath:    jobPath,
+		jobName:    name,
+		buildID:    buildID,
+	}
+}
+
+// isGCSSource recognizes whether a source string references a GCS bucket
+func isGCSSource(src string) bool {
+	return strings.HasPrefix(src, "gs://")
+}
+
+// GCSMarker holds the starting point for the next paginated GCS query
+type GCSMarker struct {
+	XMLName xml.Name `xml:"ListBucketResult"`
+	Marker  string   `xml:"NextMarker"`
+}
+
+// Contents is a single entry returned by the GCS XML API
+type Contents struct {
+	Key string
+}
+
+// GCSReq contains the contents of a GCS XML API list response
+type GCSReq struct {
+	XMLName  xml.Name `xml:"ListBucketResult"`
+	Contents []Contents
+}
+
+// names is a helper function for extracting artifact names in parallel from a GCS XML API response
+func names(content []byte, names chan string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	extracted := GCSReq{
+		Contents: []Contents{},
+	}
+	err := xml.Unmarshal(content, &extracted)
+	if err != nil {
+		logrus.WithError(err).Error("Error unmarshaling artifact names from XML")
+	}
+	for _, c := range extracted.Contents {
+		names <- c.Key
+	}
+}
+
+// tryGCSXMLGet is a helper function that attempts to list the artifacts available for the given job
+// source with the given scheme and number of retries. It returns whatever it was able to get, even if
+// it runs out of retries.
+func tryGCSXMLGet(src JobSource, scheme string, endpoint string, retries int, maxResults int) [][]byte {
+	prefix := src.JobPath()
+	bodies := [][]byte{}
+	marker := GCSMarker{}
+	c := http.Client{}
+	if scheme != "https" {
+		tr := &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+		c = http.Client{Transport: tr}
+		scheme = "http"
+	}
+	retry := 0
+	for retry < retries {
+		params := url.Values{}
+		params.Add("prefix", prefix)
+		if marker.Marker != "" {
+			params.Add("marker", marker.Marker)
+		}
+		req := url.URL{
+			Scheme:   scheme,
+			Host:     endpoint,
+			Path:     src.BucketName(),
+			RawQuery: params.Encode(),
+		}
+		resp, err := c.Get(req.String())
+		if err != nil {
+			logrus.WithError(err).Error("Error in GCS XML API GET request")
+			retry++
+			continue
+		}
+		body, err := ioutil.ReadAll(resp.Body)
+		resp.Body.Close()
+		if err != nil {
+			logrus.WithError(err).Error("Error reading body of GCS XML API response")
+			retry++
+			continue
+		}
+		bodies = append(bodies, body)
+
+		marker = GCSMarker{}
+		err = xml.Unmarshal(body, &marker)
+		if err != nil {
+			logrus.WithError(err).Error("Error unmarshaling body of GCS XML API response")
+		}
+		if marker.Marker == "" {
+			break
+		}
+	}
+	return bodies
+}
+
+// Artifacts lists all artifacts available for the given job source.
+// Uses the GCS XML API because it is ~2x faster than the golang GCS library
```

Comment:
Are we sure this is sufficient justification for rolling our own client? What does 2x faster mean?

Reply:
I think the data comes from one of the 5k-node scalability job, which lead us to switch to use this XML API cc @BenTheElder found

Reply:
@fejta 2x faster meaning listing all artifacts for a given job (i.e. with a given prefix) with the XML API is 2x faster than the go `storage` client

Reply:
The JSON APIs spit out pretty-printed data (including newlines, indent etc.), with a ton of fields we don't use. The XML API spits out only a handful of fields per entry and does the entire thing as a single line (the newlines alone are like 20k bytes per call). It's something like 1/4 of the bytes per page of 1000 results on average. For jobs with many artifacts it dropped the times from something like 5s to scan to more like 1-2s.

Reply:
The direction I gave Paul on this is to talk to the `storage` team and see if they are OK with adding client-side options to not pretty print, etc. We should fix this at the source.

Reply:
The _server_ pretty prints (curl the endpoint and see for yourself) AND the JSON API includes more fields. I don't think they're going to substantially change the JSON API for our use case in a reasonable time period. 

We can revisit this later. We're wrapping a single, simple, paginated REST / SOAP API call, from an API so stable the schema is nearly a decade old.

Reply:
The `storage` client also has good reasons for using the JSON API. And we have an unusual use case in listing a large number of keys. Normally if you needed to track lots of keys you'd do this in a separate database tracking your blobs.

Reply:
> page load time is also 2x faster

I get that 2x faster means 2x faster. What is the _impact_? Saving 5m->10m is a lot different than saving 100ms->200ms.

The scalability jobs are pretty ridiculous. I'm not 100% convinced that saving 3s when viewing this job is worth rolling our own client.

Reply:
@BenTheElder there is no way to ask the GCS server to not pretty-print? That is what I meant by editing the `storage` client

Reply:
Can't we just provide `false` to this API and save us hundreds of lines of code to maintain?

https://cloud.google.com/storage/docs/json_api/v1/parameters#pp

Reply:
Did some more digging on this for actual numbers, and yes you can send `prettyPrint=false` in the advanced params. The XML is still  _much_ smaller and faster though:

JSON, pretty print:
```console
$ curl -o $HOME/gcs-pretty.json 'https://www.googleapis.com/storage/v1/b/kubernetes-jenkins/o?prefix=logs%2Fci-kubernetes-e2e-gce-large-performance%2F121&maxResults=1000'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1256k  100 1256k    0     0  4888k      0 --:--:-- --:--:-- --:--:-- 4888k
```

JSON, don't pretty print:
```console
$ curl -o $HOME/gcs.json 'https://www.googleapis.com/storage/v1/b/kubernetes-jenkins/o?prefix=logs%2Fci-kubernetes-e2e-gce-large-performance%2F121&maxResults=1000&prettyPrint=false'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1157k  100 1157k    0     0  4865k      0 --:--:-- --:--:-- --:--:-- 4865k
```
XML:
```console
$ curl -o $HOME/gcs.xml 'https://kubernetes-jenkins.storage.googleapis.com/?prefix=logs%2Fci-kubernetes-e2e-gce-large-performance%2F121&max-keys=1000'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  330k  100  330k    0     0   959k      0 --:--:-- --:--:-- --:--:--  959k
```
---
They're ~1.2MB per page (non pretty printed JSON) vs ~330k per page (XML). Exact bytes:
```console
$ du -s -B1 $HOME/gcs.json
1187840 /usr/local/google/home/bentheelder/gcs.json

$ du -s -B1 $HOME/gcs.xml
339968  /usr/local/google/home/bentheelder/gcs.xml
```


The XML response is ~1/4 (28.62%) the size of even the non-pretty-printed JSON, because the JSON simply includes a lot more data, none of which we use. Spyglass also became subjectively much more usable for jobs with many artifacts after the switch.

Reply:
Some jobs easily have 17k artifacts (/shrug) for which we can expect roughly 20.2 MB of data vs  roughly 5.7 MB. It's definitely noticeable.

Reply:
OK, so use the `fields` parameter as exposed by the JSON API to restrict what we are asking for https://cloud.google.com/storage/docs/json_api/v1/parameters#fields

Reply:
Look, it's even smaller :)

```
curl -o $HOME/gcs.json 'https://www.googleapis.com/storage/v1/b/kubernetes-jenkins/o?prefix=logs%2Fci-kubernetes-e2e-gce-large-performance%2F121&maxResults=1000&prettyPrint=false&fields=items(id)'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  155k  100  155k    0     0   103k      0  0:00:01  0:00:01 --:--:--  103k
```

Reply:
Sure, but we'd have to roll our own client, and the XML API must remain stable due to S3 compatibility, whereas the JSON API may not, I haven't looked. This was a pretty quick change to improve the experience.